src/main/java/net/minecraftforge/common/brewing/BrewingRecipeHelper.java

```diff
@@ -0,0 +1,154 @@
+            recipeManager.getRecipeFor(ForgeMod.BREWING.get(), wrapper, level)
+                    .ifPresent(recipe -> container.set(baseSlot, recipe.assemble(wrapper)));
+        }
+    }
+
+    /**
+     * Returns true if the passed ItemStack is a valid reagent for any of the
+     * recipes in the registry.
+     */
+    public static boolean isValidIngredient(Level level, ItemStack stack)
+    {
+        if (stack.isEmpty()) return false;
+
+        for (IBrewingRecipe recipe : level.getRecipeManager().getAllRecipesFor(ForgeMod.BREWING.get()))
+        {
+            if (recipe.isReagent(stack))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the passed ItemStack is a valid base for any of the
+     * recipes in the registry.
+     */
+    public static boolean isValidInput(Level level, ItemStack stack)
+    {
+        if (stack.getCount() != 1) return false;
+
+        for (IBrewingRecipe recipe : level.getRecipeManager().getAllRecipesFor(ForgeMod.BREWING.get()))
+        {
+            if (recipe.isBase(stack))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Proxy method to reduce patch size in {@link net.minecraft.world.level.block.entity.BrewingStandBlockEntity}.
+     */
+    public static void doBrew(Level level, BlockPos pos, NonNullList<ItemStack> items)
+    {
+        if (ForgeEventFactory.onPotionAttemptBrew(items)) return;
+        ItemStack itemstack = items.get(3);
+
+        brewPotions(level, items, 3, new int[]{0, 1, 2});
+        ForgeEventFactory.onPotionBrewed(items);
+        if (itemstack.hasContainerItem())
+        {
+            ItemStack itemstack1 = itemstack.getContainerItem();
+            itemstack.shrink(1);
+            if (itemstack.isEmpty()) itemstack = itemstack1;
+            else Containers.dropItemStack(level, pos.getX(), pos.getY(), pos.getZ(), itemstack1);
+        }
+        else itemstack.shrink(1);
+
+        items.set(3, itemstack);
+        level.levelEvent(1035, pos, 0);
+    }
+
+    /**
+     * Proxy method for BrewingStandMenu.IngredientSlot#mayPlace
+     */
+    public static boolean mayPlaceIngredient(Container container, ItemStack stack)
+    {
+        if (container instanceof BrewingStandBlockEntity stand)
+            return isValidIngredient(stand.getLevel(), stack);
+        else
+            return isValidIngredient(Minecraft.getInstance().level, stack);
+    }
+
+    /**
+     * Proxy method for BrewingStandMenu.PotionSlot#mayPlaceItem
+     */
+    public static boolean mayPlacePotion(Container container, ItemStack stack)
+    {
+        if (container instanceof BrewingStandBlockEntity stand)
+            return isValidInput(stand.getLevel(), stack);
+        else
+            return isValidInput(Minecraft.getInstance().level, stack);
+    }
```

Comment:
The field is only a solution for the check in quickMoveStack because the Slot classes are static (so an additional parameter on the Slot constructor and a corresponding field would be needed)

Reply:
Thought about it some more. Probably the best thing to do in this case is use `LogicalSidedProvider` to get the client world safely. Or you could replace all level logic with `LogicalSidedProvider` if you use the forge interface for getting the current logical side.